### ISSUE TITLE
www/squid: recent segmentation faults in squid binary

### ISSUE THREAD

**Issue by Kornexl** [opnsense/plugins#3827]
**Important notices**
Before you add a new report, we ask you kindly to acknowledge the following:

- [x] I have read the contributing guide lines at https://github.com/opnsense/plugins/blob/master/CONTRIBUTING.md
- [x] I have searched the existing issues, open and closed, and I'm convinced that mine is new.
- [x] The title contains the plugin to which this issue belongs

**Describe the bug**
restarting the service produces a "Segmentation fault"

*Tip*: to validate your setup was working with the previous version, use opnsense-revert (https://docs.opnsense.org/manual/opnsense_tools.html#opnsense-revert)

**To Reproduce**
- change a value in Services - squid web proxy
- apply changes
- restart services
- a window with a last line "segmentation fault" pops up

**Expected behavior**
The service should restart with the new config; the service runs without change

**Screenshots**

**Relevant log files**
If applicable, information from log files supporting your claim.

**Additional context**
running "service squid restart" in a shell produces a similar sequence
![Screenshot 2024-02-19 190453](https://github.com/opnsense/plugins/assets/5048009/3dbaa02d-31f3-4e80-a664-db1252cc2f93)
![Screenshot 2024-02-19 191427](https://github.com/opnsense/plugins/assets/5048009/6be85a8e-ad3c-48e8-9673-a8c418f8f70f)

**Environment**
Software version used and hardware type if relevant.
e.g.:

os-squid 1.0
OPNsense 24.1.1(amd64).
on virtualbox vm

**Comment by Kornexl** [comment]
the version is os-squid 1.0 in opnsense 24.1

**Comment by Kornexl** [comment]
if i enter squid -k parse in a shell:
![Screenshot 2024-02-19 210158](https://github.com/opnsense/plugins/assets/5048009/7a2c2d67-589b-4d61-b31e-513f1418fbb6)

**Comment by dblanque** [comment]
Confirmed, have the same issue in a production environment.

**Comment by fichtner** [comment]
You can try your luck with the newer squid version 6.7 from the snapshots:

    # opnsense-revert -z squid

(run without "-z" to revert back to the stable one)


Cheers,
Franco

**Comment by dblanque** [comment]
Hi Fichtner, does this require a restart?
Installing Squid 6.7 did not affect the problem whatsoever.

**Comment by dblanque** [comment]
Update: Seems to be an issue with reloading/parsing the config file or using the squid binary whenever the service is already running. Stopping and starting the service manually seems to work fine.

**Comment by codiflow** [comment]
I can confirm that this issue still happens on 24.1.2_1.

Trying the workaroung from @dblanque brings up the following error message in the GUI but the service is being started in the background:

```
Segmentation fault
Starting squid.
Segmentation fault (core dumped)
/usr/local/etc/rc.d/squid: WARNING: failed to start squid
```

So don't be confused by the messages and always check the service status manually.

**Comment by AdSchellevis** [comment]
could be https://github.com/squid-cache/squid/commit/0129396e16573b92e7e0c81c46653f5d488968e6 , which might point to a logger issue in the configuration.

**Comment by dblanque** [comment]
Will test it today on a non-operative time-frame and get back to you on the result of the hotfix asap.
Thank you Franco.

**Comment by dblanque** [comment]
Hey Franco,
Sadly that did not fix the issue.

If there is any more info I can get for you regarding this, let me know.
Regards,
Dylan

**Comment by Dervud** [comment]
Good day.
I encountered exactly the same problem. None of the above helps.

**Comment by grimm26** [comment]
The seg fault comes from /usr/local/opnsense/scripts/proxy/setup.sh. Here is output with `set -x` from that:
```
root@edge:/etc/rc.d # service squid reload
+ SQUID_DIRS='/var/log/squid /var/run/squid /var/squid /var/squid/cache /var/squid/ssl /var/squid/logs /usr/local/etc/squid/errors/local'
+ mkdir -p /var/log/squid
+ chown -R squid:squid /var/log/squid
+ chmod -R 750 /var/log/squid
+ mkdir -p /var/run/squid
+ chown -R squid:squid /var/run/squid
+ chmod -R 750 /var/run/squid
+ mkdir -p /var/squid
+ chown -R squid:squid /var/squid
+ chmod -R 750 /var/squid
+ mkdir -p /var/squid/cache
+ chown -R squid:squid /var/squid/cache
+ chmod -R 750 /var/squid/cache
+ mkdir -p /var/squid/ssl
+ chown -R squid:squid /var/squid/ssl
+ chmod -R 750 /var/squid/ssl
+ mkdir -p /var/squid/logs
+ chown -R squid:squid /var/squid/logs
+ chmod -R 750 /var/squid/logs
+ mkdir -p /usr/local/etc/squid/errors/local
+ chown -R squid:squid /usr/local/etc/squid/errors/local
+ chmod -R 750 /usr/local/etc/squid/errors/local
+ /usr/sbin/pw groupmod proxy -m squid
+ /usr/local/sbin/squid -z -N
Segmentation fault
+ [ -f /usr/local/etc/squid/ca.pem.id ]
+ cat /usr/local/etc/squid/ca.pem.id
+ current_cert=6141ee159c02a
+ [ -d /var/squid/ssl_crtd ]
+ [ -f /var/squid/ssl_crtd.id ]
+ cat /var/squid/ssl_crtd.id
+ running_cert=6141ee159c02a
+ [ 6141ee159c02a '!=' 6141ee159c02a ]
+ [ ! -d /var/squid/ssl_crtd ]
+ /usr/local/opnsense/scripts/proxy/generate_cert.php
+ /usr/local/opnsense/scripts/proxy/deploy_error_pages.py
Performing sanity check on squid configuration.
2024/03/06 19:47:24| Processing Configuration File: /usr/local/etc/squid/squid.conf (depth 0)
2024/03/06 19:47:24| Processing Configuration File: /usr/local/etc/squid/pre-auth/40-snmp.conf (depth 1)
2024/03/06 19:47:24| Processing Configuration File: /usr/local/etc/squid/pre-auth/dummy.conf (depth 1)
2024/03/06 19:47:24| Processing Configuration File: /usr/local/etc/squid/pre-auth/parentproxy.conf (depth 1)
2024/03/06 19:47:24| Processing Configuration File: /usr/local/etc/squid/auth/dummy.conf (depth 1)
2024/03/06 19:47:24| Processing Configuration File: /usr/local/etc/squid/post-auth/dummy.conf (depth 1)
2024/03/06 19:47:24| Set Current Directory to /var/squid/cache
Segmentation fault
root@edge:/etc/rc.d # 
```

**Comment by grimm26** [comment]
> Update: Seems to be an issue with reloading/parsing the config file or using the squid binary whenever the service is already running. Stopping and starting the service manually seems to work fine.

That does not seem to be the case. I can fully stop squid and do something simple like `squid -k check` and get a seg fault. It drops a `squid.core` in `/var/squid` so I ran a strings on that and I do see a
 ```
Invalid shared object handle 0x82ad391b0
__vdso_ia32_sigcode
__vdso_freebsd4_ia32_sigcode
elf-vdso32.so.1
```
 towards the end of it. Not sure if those lines being together means anything. I was going to compile squid myself and see if that also threw an error, but I'm not a FreeBSD guy and the make failed.

**Comment by fichtner** [comment]
I just uploaded a squid build for 6.8. setup.sh or not if squid binary crashes we can't do that much I'm afraid. But I'm repeating myself.

To install the 6.8 version:

    # opnsense-revert -z squid

Best to test with a clean reboot afterwards.


Cheers,
Franco

**Comment by grimm26** [comment]
@fichtner squid 6.8 still segfaults for me like 6.7 did.

**Comment by fichtner** [comment]
Fair enough. Was worth a try. ;)

**Comment by gridstop** [comment]
I've had the exact same issue, so I was trying a lot of things, and I think one of these two things fixed it:
A) In the general settings, I changed the error page dropdown from OPNsense -> Squid
B) I then unchcked the 'Proxy enabled' and applied, and then rechecked and applied, and it started correctly.
This was after trying to restart multiple times, including from command line, all giving segfaults. I'm not sure which of these two was actually the key, but maybe this points someone else in the right direction.

**Comment by grimm26** [comment]
@gridstop I did that and it did not help. Still get a seg fault.

**Comment by pablopol24** [comment]
> I've had the exact same issue, so I was trying a lot of things, and I think one of these two things fixed it: A) In the general settings, I changed the error page dropdown from OPNsense -> Squid B) I then unchcked the 'Proxy enabled' and applied, and then rechecked and applied, and it started correctly. This was after trying to restart multiple times, including from command line, all giving segfaults. I'm not sure which of these two was actually the key, but maybe this points someone else in the right direction.

it didn't work either :(

**Comment by gridstop** [comment]
Well, sorry about that and nevermind then. I looked through the config again, I genuinely can't see anything else I changed, and it just suddenly started working when I unchecked & rechecked the enable box. It had never started successfully since upgrading to 24.1, including not starting on reboots.

**Comment by wiigoac** [comment]
I have affected installations too. Nothing fancy. Just os-squid and even removed anything left of older squid installations from the system and from config.xml. Still I observe this:

<pre>
root:/usr/local/etc/squid # squid -d9 -f squid.conf -k parse                                                                                                                                               2024/03/18 12:04:29| Processing Configuration File: squid.conf (depth 0)
2024/03/18 12:04:29| Processing: http_port 10.0.0.100:3128
2024/03/18 12:04:29| Processing: acl ftp proto FTP
2024/03/18 12:04:29| Processing: http_access allow ftp
2024/03/18 12:04:29| Processing: acl localnet src 10.63.36.0/23 # Possible internal network (interfaces v4)
2024/03/18 12:04:29| Processing: acl localnet src fc00::/7       # RFC 4193 local private network range
2024/03/18 12:04:29| Processing: acl localnet src fe80::/10      # RFC 4291 link-local (directly plugged) machines
2024/03/18 12:04:29| Processing: acl unrestricted src 10.0.0.1
2024/03/18 12:04:29| Processing: acl unrestricted src 10.0.0.2
2024/03/18 12:04:29| Processing: acl unrestricted src 10.0.0.3
2024/03/18 12:04:29| Processing: acl unrestricted src 10.0.0.4
2024/03/18 12:04:29| Processing: acl SSL_ports port 443 # https
2024/03/18 12:04:29| Processing: acl Safe_ports port 80 # http
2024/03/18 12:04:29| Processing: acl Safe_ports port 443 # https
2024/03/18 12:04:29| Processing: acl CONNECT method CONNECT
2024/03/18 12:04:29| Processing: icap_enable off
2024/03/18 12:04:29| Processing: include /usr/local/etc/squid/pre-auth/*.conf
2024/03/18 12:04:29| Processing Configuration File: /usr/local/etc/squid/pre-auth/40-snmp.conf (depth 1)
2024/03/18 12:04:29| Processing Configuration File: /usr/local/etc/squid/pre-auth/dummy.conf (depth 1)
2024/03/18 12:04:29| Processing Configuration File: /usr/local/etc/squid/pre-auth/parentproxy.conf (depth 1)
2024/03/18 12:04:29| Processing: http_access allow unrestricted
2024/03/18 12:04:29| Processing: http_access deny !Safe_ports !unrestricted
2024/03/18 12:04:29| Processing: http_access deny CONNECT !SSL_ports !unrestricted
2024/03/18 12:04:29| Processing: http_access allow localhost manager
2024/03/18 12:04:29| Processing: http_access deny manager
2024/03/18 12:04:29| Processing: http_access deny to_localhost
2024/03/18 12:04:29| Processing: include /usr/local/etc/squid/auth/*.conf
2024/03/18 12:04:29| Processing Configuration File: /usr/local/etc/squid/auth/dummy.conf (depth 1)
2024/03/18 12:04:29| Processing: http_access allow localnet
2024/03/18 12:04:29| Processing: http_access allow localhost
2024/03/18 12:04:29| Processing: http_access deny all
2024/03/18 12:04:29| Processing: include /usr/local/etc/squid/post-auth/*.conf
2024/03/18 12:04:29| Processing Configuration File: /usr/local/etc/squid/post-auth/dummy.conf (depth 1)
2024/03/18 12:04:29| Processing: cache_mem 256 MB
2024/03/18 12:04:29| Processing: coredump_dir /var/squid/cache
2024/03/18 12:04:29| Processing: refresh_pattern ^ftp:          1440    20%     10080
2024/03/18 12:04:29| Processing: refresh_pattern ^gopher:       1440    0%      1440
2024/03/18 12:04:29| Processing: refresh_pattern -i (/cgi-bin/|\?) 0    0%      0
2024/03/18 12:04:29| Processing: refresh_pattern .              0       20%     4320
2024/03/18 12:04:29| Processing: pinger_enable off
2024/03/18 12:04:29| Processing: access_log stdio:/var/log/squid/access.log squid
2024/03/18 12:04:29| Processing: cache_store_log stdio:/var/log/squid/store.log
2024/03/18 12:04:29| Processing: logfile_rotate 0
2024/03/18 12:04:29| Processing: cache_mgr admins
2024/03/18 12:04:29| Processing: error_directory /usr/local/share/squid-langpack/en
2024/03/18 12:04:29| Requiring client certificates.
Segmentation fault
root:/usr/local/etc/squid #                                     
</pre>

I

**Comment by wiigoac** [comment]
With a little bit more playing around. I would say it has to be more the "checking of config" part than the actual "running proxy" part of the story. As was mentioned before. You can get this:

![image](https://github.com/opnsense/plugins/assets/64070864/cb4f8fa9-934e-4b31-a96e-d35fd9387c04)

and have a running squid afterwards. Also I can run the process manual in foreground and seems to be ok, while segfaulting on the "-k parse".
I would love to give more secific information, but I am afraid I'll not have too much time to invest. Although, if there is something to try out, I will do so, of course.

**Comment by ssteeltm** [comment]
it also happens on mine, even with # opnsense-revert -z squid 6.8

**Comment by dblanque** [comment]
Oof, since 24.1.4 we've been getting intermittent crashes on squid. There's something crooked with Squid on the new version.

**Comment by densu90** [comment]
+1
i have the:
```
Segmentation fault
Performing sanity check on squid configuration.
2024/03/21 17:38:22| Processing Configuration File: /usr/local/etc/squid/squid.conf (depth 0)
2024/03/21 17:38:22| Starting Authentication on port 127.0.0.1:3128
2024/03/21 17:38:22| Disabling Authentication on port 127.0.0.1:3128 (interception enabled)
2024/03/21 17:38:22| Starting Authentication on port [::1]:3128
2024/03/21 17:38:22| Disabling Authentication on port [::1]:3128 (interception enabled)
2024/03/21 17:38:22| Starting Authentication on port 127.0.0.1:3129
2024/03/21 17:38:22| Disabling Authentication on port 127.0.0.1:3129 (interception enabled)
2024/03/21 17:38:22| Starting Authentication on port [::1]:3129
2024/03/21 17:38:22| Disabling Authentication on port [::1]:3129 (interception enabled)
2024/03/21 17:38:22| Processing Configuration File: /usr/local/etc/squid/pre-auth/40-snmp.conf (depth 1)
2024/03/21 17:38:22| Processing Configuration File: /usr/local/etc/squid/pre-auth/dummy.conf (depth 1)
2024/03/21 17:38:22| Processing Configuration File: /usr/local/etc/squid/pre-auth/parentproxy.conf (depth 1)
2024/03/21 17:38:22| Processing Configuration File: /usr/local/etc/squid/auth/dummy.conf (depth 1)
2024/03/21 17:38:22| Processing Configuration File: /usr/local/etc/squid/post-auth/dummy.conf (depth 1)
2024/03/21 17:38:22| WARNING: HTTP requires the use of Via
2024/03/21 17:38:23| Set Current Directory to /var/squid/cache
Segmentation fault
```
on service restart and
```
Segmentation fault
Starting squid.
Segmentation fault (core dumped)
/usr/local/etc/rc.d/squid: WARNING: failed to start squid
```
on service start


in addition, i need to wait 5 minutes after Opnsense started before squid starts to work properly. before that i will get no error page from squid if try reach any web site

_PS: a had reinstalled clean and reconfigure Opnsense by hand (not loading settings) on new VM (empty volume) in struggle to get rid of the issue. unsuccessfully._

**Comment by AdSchellevis** [comment]
To exclude configuration issues in our templates, I tried to execute squid using the `/usr/local/etc/squid/squid.conf.sample` file as well, which also seems to coredump. At a first glance it dies on initializing the client tls context. When disabling `tls_outgoing_options`, `squid -k parse` exists without coredump.

```
tls_outgoing_options disable
```

When increasing debug output, I instantly hit the next coredump.... which also makes it impossible to trace errors inside the tls_outgoing_options block as it will never reach there with `-X` appended (on my end)

**Comment by fichtner** [comment]
ASLR related perhaps?

**Comment by AdSchellevis** [comment]
I was thinking the same, but disabling aslr (`kern.elf64.aslr.enable=0`) doesn't make a difference.

**Comment by batonac** [comment]
Interestingly enough, restarting the service from the Dashboard Services widget works fine for me, but restarting from the Squid Web Proxy Service UI gives the segmentation fault.

**Comment by densu90** [comment]
> Interestingly enough, restarting the service from the Dashboard Services widget works fine for me, but restarting from the Squid Web Proxy Service UI gives the segmentation fault.

i'll rather believe in service messages suppression on main page

**Comment by sharifm-informatica** [comment]
The same issue. Segmentation fault (core dumped) still happens on OPNsense 24.1.6 and with squid 6.9. Last few lins of result of `squid -k parse`:

```
2024/04/18 13:36:16| Processing: logfile_rotate 0
2024/04/18 13:36:16| Processing: cache_mgr admin@localhost.local
2024/04/18 13:36:16| Processing: error_directory /usr/local/etc/squid/errors/local
2024/04/18 13:36:16| WARNING: use of 'reload-into-ims' in 'refresh_pattern' violates HTTP
2024/04/18 13:36:16| Requiring client certificates.
Segmentation fault (core dumped)
```

**Comment by lisachev** [comment]
Strange enough, the 'Apply' button does not invoke any config reload for Squid. Say adding a blacklist host and hitting Apply will not immediately apply the config. Service restart attempt will fail with segfault. Rebooting the whole OPNsense will go fine and the new config will be applied as well.  Every 'Apply' press logs an 'returned exit status 1' error from config.py.
There seems to be something broken with the config checker..

**Comment by davidemiccone** [comment]
Still some problem here 24.1.6 squid 6.9
Anyone know if there is some plan to resolve?

**Comment by 0x646f6d** [comment]
Hi there!

Same issue here while using brand new opnsense business edition 24.4 with Squid version 6.9.
I thought business edition is supposed to be serving more stable software?!

I can confirm, that this is not related to the configuration, running squid with a minimal fresh config also segfaults.

No matter how I start or restart squid (GUI oder commandline), it seems that some parts of squid still start up, as the processes and also TCP Port 8080 come up. 

```
root@redacted:~ # ps aux | grep squid
squid   97828   0.0  0.1   21292   10336  -  S    09:47      0:00.11 (pinger) (pinger)
root     3943   0.0  0.0   12720    2244  1  S+   10:30      0:00.00 grep squid

root@redacted:~ # netstat -na -f inet | grep LISTEN
tcp4       0      0 127.0.0.1.6379         *.*                    LISTEN
tcp4       0      0 *.443                  *.*                    LISTEN
tcp4       0      0 *.80                   *.*                    LISTEN
tcp4       0      0 127.0.0.1.43580        *.*                    LISTEN
tcp4       0      0 127.0.0.1.953          *.*                    LISTEN
tcp4       0      0 *.53                   *.*                    LISTEN
tcp4       0      0 *.8443                 *.*                    LISTEN
tcp4       0      0 *.22                   *.*                    LISTEN

root@redacted:~ # configctl proxy start
Segmentation fault
Starting squid.
Segmentation fault (core dumped)
/usr/local/etc/rc.d/squid: WARNING: failed to start squid

root@redacted:~ # ps aux | grep squid
squid   16301   5.7  0.5  640832   91432  -  S    10:30      0:00.66 (squid-1) --kid squid-1 -f /usr/local/etc/squid/squid.conf (squid)
squid   20405   5.6  0.1   21292   10320  -  S    10:30      0:00.05 (pinger) (pinger)
squid   20137   4.4  0.0   16108    4928  -  S    10:30      0:00.04 (unlinkd) (unlinkd)
squid   14664   1.9  0.1  302420   20552  -  Ss   10:30      0:00.00 /usr/local/sbin/squid -f /usr/local/etc/squid/squid.conf
squid    4769   0.0  0.1   21292   10328  -  S    09:28      0:00.13 (pinger) (pinger)
squid   97828   0.0  0.1   21292   10336  -  S    09:47      0:00.12 (pinger) (pinger)
root    70177   0.0  0.0   12720    2252  1  S+   10:30      0:00.00 grep squid

root@redacted:~ # netstat -na -f inet | grep LISTEN
tcp46      0      0 *.8080                 *.*                    LISTEN
tcp4       0      0 10.0.0.1.8080          *.*                    LISTEN
tcp4       0      0 127.0.0.1.6379         *.*                    LISTEN
tcp4       0      0 *.443                  *.*                    LISTEN
tcp4       0      0 *.80                   *.*                    LISTEN
tcp4       0      0 127.0.0.1.43580        *.*                    LISTEN
tcp4       0      0 127.0.0.1.953          *.*                    LISTEN
tcp4       0      0 *.53                   *.*                    LISTEN
tcp4       0      0 *.53                   *.*                    LISTEN
tcp4       0      0 *.8443                 *.*                    LISTEN
tcp4       0      0 *.22                   *.*                    LISTEN
```

I confirmed the proxy is working on another machine.

For me it seems to be an issue with the squid port for FreeBSD...

**Comment by fichtner** [comment]
Squid and OPNProxy have been demoted to community plugins because of security and release quality concerns. What you see here is the merely the reason for this decision.


Cheers,
Franco

**Comment by densu90** [comment]
i don't know what changed exactly, but it seems to be working for me now. at least 4 hours and flight is good. i didn't reboot yet, but still. it still throwing errors when try to restart servise, but otherwise it works fine. with no config changes.
OPNsense 24.1.7_4-amd64
Thx!
ps: please, do not brick it again if possible ;)

**Comment by fichtner** [comment]
> ps: please, do not brick it again if possible ;)

I know it's meant as funny but this was a hard nut to crack involving OpenSSL 3, broken legacy compatibility in other parts of the system and squid for one reason or another crashing due to working around other parts of the system breaking.

I just found the underlying problem by accident talking about it in the forum with a user.

**Comment by Kornexl** [comment]
I updated my instance to the latest version yesterday but the segmentation fault occours at the start of squid. Where do i find the corrected version of squid for opnsense

**Comment by fichtner** [comment]
You probably need to reboot just to be sure.

**Comment by Kornexl** [comment]
Today i updated again und the update to 24.1.8 rebooted the system. 
But the segmentation fault ist still there. The squid process runs, but the segmentation fault occurs every time i restart the service squid.
Should i open a new issue or can this issue be reopened

![Screenshot 2024-06-06 102728](https://github.com/opnsense/plugins/assets/5048009/570468a9-e92e-4d87-8aff-c6c8755d9595)

![Screenshot 2024-06-06 102033](https://github.com/opnsense/plugins/assets/5048009/a840d324-34d5-4a86-9987-98ab2925c108)
![Screenshot 2024-06-06 102515](https://github.com/opnsense/plugins/assets/5048009/d1416260-1eff-4a40-a1a0-a6638b5318f3)

**Comment by densu90** [comment]
confirm, now its broken again. 24.1.8. it just faling apart after some time after sqid started and need sqid stop and sqid start to continue. sqid restart for last year never worked correctly. errors in place, but it was in 24.1.7_4 too, yet in 24.1.7_4 whole setup worked without bricking while 24.1.8 it's not.

**Comment by fichtner** [comment]
Ok, technically, how would that be possible?

**Comment by dblanque** [comment]
I can also confirm that the service does produce a segmentation fault on service restart as of 24.1.8 but does NOT produce it on stop-start from the dashboard.

**Comment by jolly-jump** [comment]
Hi. 
Not sure, if related, but I am on 24.1.8 and now on 24.1.9 and I see the segfault on the command line:
```
# /usr/local/etc/rc.d/squid forcerestart
Segmentation fault
Performing sanity check on squid configuration.
2024/06/18 19:06:27| Processing Configuration File: /usr/local/etc/squid/squid.conf (depth 0)
2024/06/18 19:06:27| Processing Configuration File: /usr/local/etc/squid/pre-auth/20-negotiate.auth.conf (depth 1)
2024/06/18 19:06:27| Processing Configuration File: /usr/local/etc/squid/pre-auth/40-snmp.conf (depth 1)
2024/06/18 19:06:27| Processing Configuration File: /usr/local/etc/squid/pre-auth/50-linuxmuster.pre-auth.conf (depth 1)
2024/06/18 19:06:27| Processing Configuration File: /usr/local/etc/squid/pre-auth/dummy.conf (depth 1)
2024/06/18 19:06:27| Processing Configuration File: /usr/local/etc/squid/pre-auth/parentproxy.conf (depth 1)
2024/06/18 19:06:27| Processing Configuration File: /usr/local/etc/squid/auth/dummy.conf (depth 1)
2024/06/18 19:06:27| Processing Configuration File: /usr/local/etc/squid/post-auth/dummy.conf (depth 1)
2024/06/18 19:06:27| Set Current Directory to /var/squid/cache
Segmentation fault
Performing sanity check on squid configuration.
2024/06/18 19:06:27| Processing Configuration File: /usr/local/etc/squid/squid.conf (depth 0)
2024/06/18 19:06:27| Processing Configuration File: /usr/local/etc/squid/pre-auth/20-negotiate.auth.conf (depth 1)
2024/06/18 19:06:27| Processing Configuration File: /usr/local/etc/squid/pre-auth/40-snmp.conf (depth 1)
2024/06/18 19:06:27| Processing Configuration File: /usr/local/etc/squid/pre-auth/50-linuxmuster.pre-auth.conf (depth 1)
2024/06/18 19:06:27| Processing Configuration File: /usr/local/etc/squid/pre-auth/dummy.conf (depth 1)
2024/06/18 19:06:27| Processing Configuration File: /usr/local/etc/squid/pre-auth/parentproxy.conf (depth 1)
2024/06/18 19:06:27| Processing Configuration File: /usr/local/etc/squid/auth/dummy.conf (depth 1)
2024/06/18 19:06:27| Processing Configuration File: /usr/local/etc/squid/post-auth/dummy.conf (depth 1)
2024/06/18 19:06:27| Set Current Directory to /var/squid/cache
Segmentation fault
Stopping squid.
Waiting for PIDS: 61243.
Starting squid.
```
What bothers me is when I reboot the squid starts and quits. The cache.log says:
```
2024/06/18 07:00:46 kid1| helperOpenServers: Starting 0/5 'ext_kerberos_ldap_group_acl' processes
2024/06/18 07:00:46 kid1| helperOpenServers: No 'ext_kerberos_ldap_group_acl' processes needed.
2024/06/18 07:00:47 kid1| ERROR: loading file '/usr/local/etc/squid/errors/local/error-details.txt': (2) No such file or directory
2024/06/18 07:00:47 kid1| Logfile: opening log stdio:/var/log/squid/access.log
2024/06/18 07:00:47 kid1| Local cache digest enabled; rebuild/rewrite every 3600/3600 sec
2024/06/18 07:00:47 kid1| Logfile: opening log stdio:/var/log/squid/store.log
2024/06/18 07:00:47 kid1| Swap maxSize 0 + 262144 KB, estimated 20164 objects
2024/06/18 07:00:47 kid1| Target number of buckets: 1008
2024/06/18 07:00:47 kid1| Using 8192 Store buckets
2024/06/18 07:00:47 kid1| Max Mem  size: 262144 KB
2024/06/18 07:00:47 kid1| Max Swap size: 0 KB
2024/06/18 07:00:47 kid1| Using Least Load store dir selection
2024/06/18 07:00:47 kid1| Set Current Directory to /var/squid/cache
2024/06/18 07:00:47 kid1| Finished loading MIME types and icons.
2024/06/18 07:00:47 kid1| HTCP Disabled.
2024/06/18 07:00:47 kid1| sendto FD 12: (13) Permission denied
2024/06/18 07:00:47 kid1| ERROR: ipcCreate: CHILD: hello write test failed
2024/06/18 09:18:28| Removing PID file (/var/run/squid/squid.pid)
```
and when I force restart the service, it works just fine (beside the said segfaults):
```
2024/06/18 09:18:28 kid1| ERROR: loading file '/usr/local/etc/squid/errors/local/error-details.txt': (2) No such file or directory
2024/06/18 09:18:28 kid1| Logfile: opening log stdio:/var/log/squid/access.log
2024/06/18 09:18:28 kid1| Local cache digest enabled; rebuild/rewrite every 3600/3600 sec
2024/06/18 09:18:28 kid1| Logfile: opening log stdio:/var/log/squid/store.log
2024/06/18 09:18:28 kid1| Swap maxSize 0 + 262144 KB, estimated 20164 objects
2024/06/18 09:18:28 kid1| Target number of buckets: 1008
2024/06/18 09:18:28 kid1| Using 8192 Store buckets
2024/06/18 09:18:28 kid1| Max Mem  size: 262144 KB
2024/06/18 09:18:28 kid1| Max Swap size: 0 KB
2024/06/18 09:18:28 kid1| Using Least Load store dir selection
2024/06/18 09:18:28 kid1| Set Current Directory to /var/squid/cache
2024/06/18 09:18:28 kid1| Finished loading MIME types and icons.
2024/06/18 09:18:28 kid1| HTCP Disabled.
2024/06/18 09:18:28 kid1| Pinger socket opened on FD 13
2024/06/18 09:18:28 kid1| Squid plugin modules loaded: 0
2024/06/18 09:18:28 kid1| Adaptation support is off.
2024/06/18 09:18:28 kid1| Accepting HTTP Socket connections at conn3 local=10.16.1.254:3128 remote=[::] FD 11 flags=9
```

if this is failure of "fd: 12 (13)" is not related to this segfault, then sorry for the wrongdoing and please redirect me, if there is another place to report or open up a separate issue.

Thanks.

**Comment by Kornexl** [comment]
I have tested the installations 23.x and 24.x again. 
I found that the segmentation fault appears in 23.x too, but no alert in the dashboard pops up (no proxy load error message).

I have started both squid (6.06 and 6.10) with debug_options ALL,5 


I have found the following lines in cache.log (grep _suid cache.log)

2024/07/22 17:26:52.186 kid1| 21,3| tools.cc(625) enter_suid: enter_suid: PID 29145 taking root privileges
2024/07/22 17:26:52.186 kid1| 21,31 tools.cc(629) enter_suid: enter_suid: setresuid failed: (1) Operation not permitted
2024/07/22 17:26:52.186 kid1| 21,3| tools.cc(561) leave suid: leave suid: PID 29145 called
2024/07/22 17:26:52.187 kid1| 21,31 tools.cc(625) enter_suid: enter_suid: PID 29145 taking root privileges
2024/07/22 17:26:52.187 kid1| 21,3| tools.cc(629) enter_suid: enter_suid: setresuid failed: (1) Operation not permitted
2024/07/22 17:26:52.187 kid1| 21,31 tools.cc(561) leave_suid: leave suid: PID 29145 called
2024/07/22 17:26:52.187 kid1| 21,31 tools.cc(625) enter_suid: enter_suid: PID 29145 taking root privileges
2024/07/22 17:26:52.187 kid1l 21,31 tools.cc(629) enter_suid: enter_suid: setresuid failed: (1) Operation not permitted
2024/07/22 17:26:52.187 kid1| 21,31 tools.cc(561) leave_suid: leave suid: PID 29145 called
2024/07/22 17:26:52.187 kid1l 21,31 tools.cc(561) leave_suid: leave_suid: PID 29648 called
2024/07/22 17:26:52.187 kid1l 21,31 tools.cc(651) no_suid: no_suid: PID 29648 giving up root privileges forever

maybe this "operation not permitted" is the cause of the "segmentation fault".

The difference between the installations 23.x and 24.x is the alerting of this segmentation fault in the dashboard of opnsense.

But what ist the cause of this "Operation not permitted"

yours

Anton Kornexl

**Comment by Kornexl** [comment]
I have created a bug report  in bugs.squid-cache.org: https://bugs.squid-cache.org/show_bug.cgi?id=5390
The answer there was:
--------------------
I can see that exiting Squid is crashing while trying to clean its TLS configuration state. There is a known problem in how Squid stores a part of that state (I have updated this bug title accordingly). That problem does not always result in segmentation faults in an exiting Squid process, but I speculate that it does in your environment.
-----------------
Maybe we get a corrected squid binary some day.

The  "operation not permitted" didn't cause the Segmentation fault.

**Comment by Kornexl** [comment]
The bug in squid (https://bugs.squid-cache.org/show_bug.cgi?id=5390) is fixed, but only in the upcoming squid v7 (no backport to v6).

**Comment by Vivekmauli14** [comment]
Hopefully the long awaited issue will be resolved soon. Looking forward for the update, in the meantime can anyone guide me how can I make squid working on my 24.7 version ?

**Comment by J-Damm** [comment]
I had success with creating a internal CA and select it for the Squid TLS inspection...

**Comment by montjoie** [comment]
This bug is still here on 24.7.12_2

**Comment by cpeghkz221** [comment]
The fix appears to have landed in commit 9e779e4 of the master/v8 branch.  Would it be possible to get an update on this?

**Comment by davidemiccone** [comment]
> The fix appears to have landed in commit 9e779e4 of the master/v8 branch.  Would it be possible to get an update on this?

master/v8 for squid?
Where you found this?
